### PR TITLE
Fix incorrect sln add generation

### LINKED ISSUE #16137: dotnet sln add command is not checking for project name conflicts.
## Steps to reproduce

With existing project name conflicting with an added solution folder:

```
$ mkdir tmp
$ cd tmp
$ mkdir foo
$ cd foo
$ dotnet new lib
$ mkdir bar
$ cd bar
$ dotnet new lib
$ cd ../..
$ dotnet new sln
$ dotnet sln add foo
$ dotnet sln add foo/bar
$ dotnet build
```

or with an existing solution folder that conflicts with a project of the same name being added:

```
$ mkdir tmp
$ cd tmp
$ mkdir foo
$ cd foo
$ dotnet new lib
$ mkdir bar
$ cd bar
$ dotnet new lib
$ cd ../..
$ dotnet new sln
$ dotnet sln add foo/bar
$ dotnet sln add foo
$ dotnet build
```
## Expected  behavior

The solution builds successfully.

## Actual behavior

The build fails.

```
$ dotnet build
Microsoft (R) Build Engine version 15.7.66.2115 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

/Users/peterhuene/repro/repro.sln : Solution file error MSB5004: The solution file has two projects named "foo".

Build FAILED.

/Users/peterhuene/repro/repro.sln : Solution file error MSB5004: The solution file has two projects named "foo".
    0 Warning(s)
    1 Error(s)
```

## Environment data
```
$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.1.300-preview3-008413
 Commit:    112926727a

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.13
 OS Platform: Darwin
 RID:         osx.10.13-x64
 Base Path:   /Users/peterhuene/src/cli/bin/2/osx-x64/dotnet/sdk/2.1.300-preview3-008413/

Host (useful for support):
  Version: 2.1.0-preview2-26314-02
  Commit:  8ac5f4fa22

.NET Core SDKs installed:
  2.1.300-preview3-008413 [/Users/peterhuene/src/cli/bin/2/osx-x64/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0-preview2-30338 [/Users/peterhuene/src/cli/bin/2/osx-x64/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0-preview2-30338 [/Users/peterhuene/src/cli/bin/2/osx-x64/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.0-preview2-26314-02 [/Users/peterhuene/src/cli/bin/2/osx-x64/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

### PR DESCRIPTION
**Pull request by mfkl**
Fixes #16137

Hi,

This is a PR to try to address https://github.com/dotnet/sdk/issues/9220

I wrote a test case which reproduces the issue in https://github.com/dotnet/sdk/commit/385d3ead1c4ce17537b1bb6107a65bb75d5c8ba3. Instead of checking the SLN for an expected line-by-line output, I try to build it to make sure it is valid. I'm aware this is crossing the scope of your `dotnet build` tests, but I thought this kind of "deeper" integration test could also be useful, IMHO.

I'm not too sure about the following patch https://github.com/dotnet/sdk/commit/b10a9af2ce5dcab2cb3fb6a5527c3f3473273d3a and would welcome your feedback. It is kind of a workaround, but gets the job done to address https://github.com/dotnet/sdk/issues/9220.

I tried looking at how https://github.com/dotnet/project-system does it but could not find much info with regard to `NestedProjects`. Also failed to generate that node from VS2019 by nesting various projects in a solution.

Let me know what you think.

### CODE REVIEW

**Comment by mfkl** [comment]
ping :-)

/cc @dsplaisted @joeloff @marcpopMSFT (not sure who owns this)

**Comment by mfkl** [comment]
ping ^

**Comment by mfkl** [comment]
Some of the tests are failing because the generated SLN is different following https://github.com/dotnet/sdk/pull/11846/commits/7754a88211818e66bbc4e9942c57d770a485b13e. 

I believe 
```
private static IList<string> GetSolutionFoldersFromProjectPath(string projectFilePath)
{
    var solutionFolders = new List<string>();

    if (!IsPathInTreeRootedAtSolutionDirectory(projectFilePath))
        return solutionFolders;

    var currentDirString = $".{Path.DirectorySeparatorChar}";
    if (projectFilePath.StartsWith(currentDirString))
    {
        projectFilePath = projectFilePath.Substring(currentDirString.Length);
    }

    var projectDirectoryPath = TrimProject(projectFilePath);
    if (string.IsNullOrEmpty(projectDirectoryPath))
        return solutionFolders;

    var solutionFoldersPath = TrimProjectDirectory(projectDirectoryPath);
    if (string.IsNullOrEmpty(solutionFoldersPath))
        return solutionFolders;

    solutionFolders.AddRange(solutionFoldersPath.Split(Path.DirectorySeparatorChar));

    return solutionFolders;
}
```
doesn't take into account all the project nesting possible cases that various solutions can have.

**Comment by stackedsax** [comment]
hey there @joeloff, any thoughts on this one?

**Comment by marcpopMSFT** [comment]
Coming back to old PRs. The original fix didn't quite work right as it wouldn't handle the different add order and the solution folder path wasn't quite right.  With the new fix, detect if there is an existing project that matches the solution folder about to be created and if so, add that project to it's own solution folder. If the add is just a project, check if there is a duplicate and add it to it's own solution folder if needed.  This fix won't handle if someone is adding duplicate project within the same solution folder hierarchy but it's closer.

File: src/Cli/dotnet/SlnFileExtensions.cs [hunk 1]
```diff
@@ -222,6 +243,11 @@ private static void AddSolutionFolders(this SlnFile slnFile, SlnProject slnProje
 
                 var pathToGuidMap = slnFile.GetSolutionFolderPaths(nestedProjectsSection.Properties);
 
+                if (slnFile.HasSolutionFolder(nestedProjectsSection.Properties, slnProject))
+                {
```
Comment (nagilson) [inline src/Cli/dotnet/SlnFileExtensions.cs:247]: Should we also add some sort of check with an error if the folder has a project with the desired name?

File: src/Cli/dotnet/SlnFileExtensions.cs [hunk 2]
```diff
@@ -83,6 +83,27 @@ public static void AddProject(this SlnFile slnFile, string fullProjectPath, ILis
 
                 if (solutionFolders != null)
                 {
```
Comment (nagilson) [inline src/Cli/dotnet/SlnFileExtensions.cs:85]: With this added code, this scope-block becomes complex enough that it should be delegated into it's own set of logic, especially considering it is easily scopeable to a single "thing." I would move it into a function called `SetupSolutionFolders`, `ValidateSolutionFolderLogic`, or something like that.

File: src/Cli/dotnet/SlnFileExtensions.cs [hunk 3]
```diff
@@ -83,6 +83,27 @@ public static void AddProject(this SlnFile slnFile, string fullProjectPath, ILis
...
+                        // If the project conflicts with a solution folder, add it's own folder as a solution folder
+                        var duplicateProject = slnFile.Projects.Where(p => string.Equals(p.Name, slnProject.Name, StringComparison.OrdinalIgnoreCase)
+                                                                       && p.TypeGuid == ProjectTypeGuids.SolutionFolderGuid).FirstOrDefault();
+                        if (duplicateProject != null)
+                        {
+                            slnFile.AddSolutionFolders(slnProject, new List<string>() { Path.GetDirectoryName(relativeProjectPath) });
+                        }
+                    }
+
                     slnFile.AddSolutionFolders(slnProject, solutionFolders);
```
Comment (nagilson) [inline src/Cli/dotnet/SlnFileExtensions.cs:107]: I was confused at first as to why we did this line even if we added to the folder above, so I think this is helpful
```suggestion
                   // Even if we added a solution folder above, we still need to add the folder specified by the command option.
                    slnFile.AddSolutionFolders(slnProject, solutionFolders);
```

File: src/Cli/dotnet/SlnFileExtensions.cs [hunk 4]
```diff
@@ -83,6 +83,27 @@ public static void AddProject(this SlnFile slnFile, string fullProjectPath, ILis
...
+                        var duplicateProjects = slnFile.Projects.Where(p => solutionFolders.Contains(p.Name)
+                                                                        && p.TypeGuid != ProjectTypeGuids.SolutionFolderGuid).ToList();
+                        foreach (SlnProject duplicateProject in duplicateProjects)
+                        {
+                            slnFile.AddSolutionFolders(duplicateProject, new List<string>() { Path.GetDirectoryName(duplicateProject.FilePath) });
+                        }
+                    }
+                    else
+                    {
+                        // If the project conflicts with a solution folder, add it's own folder as a solution folder
```
Comment (nagilson) [inline src/Cli/dotnet/SlnFileExtensions.cs:98]: > add it's own folder as a solution folder

What does this mean? I am confused by this comment
```suggestion
                        // If the project has the same name of a solution folder, give it a solution folder named ??? to separate the two 
```

File: src/Cli/dotnet/SlnFileExtensions.cs [hunk 5]
```diff
@@ -83,6 +83,27 @@ public static void AddProject(this SlnFile slnFile, string fullProjectPath, ILis
 
                 if (solutionFolders != null)
                 {
+                    if (solutionFolders.Any())
+                    {
+                        // Check for any existing projects if there is a new matching solution folder and add them to their own solution folder
```
Comment (nagilson) [inline src/Cli/dotnet/SlnFileExtensions.cs:88]: I think "Project" can also mean "Solution Folder" so this is confusing me. Is it trying to say: If the solution folder to add has the same name of an existing solution folder then add it to the existing folder? I don't think so.

**Review by nagilson — COMMENTED** [review]
Please see the requested changes @marcpopMSFT, thanks for working on this PR. In addition, thank you to @mfkl for making this community contribution, it's a very nice bug fix to have. I'm sorry it took us a while to get to it. I also happen to love VLC and use it a lot, so I appreciate your work 😄

File: src/Cli/dotnet/SlnFileExtensions.cs (hunk 1 above)
Comment (marcpopMSFT) [reply src/Cli/dotnet/SlnFileExtensions.cs:253]: I added the error though the logic was a bit more complex that I'd like. Take a look.

Comment (nagilson) [reply src/Cli/dotnet/SlnFileExtensions.cs:253]: Awesome, thank you :)

File: src/Cli/dotnet/SlnFileExtensions.cs [hunk 6]
```diff
@@ -93,6 +90,36 @@ public static void AddProject(this SlnFile slnFile, string fullProjectPath, ILis
...
+                }
+                else
+                {
+                    // If a project and solution folder have the same name, add it's own folder as a solution folder
+                    // eg. foo\extensions.csproj and extensions\library\library.csproj would have a project and solution folder with conflicting names
+                    var duplicateProject = slnFile.Projects.Where(p => string.Equals(p.Name, slnProject.Name, StringComparison.OrdinalIgnoreCase)
+                                                                   && p.TypeGuid == ProjectTypeGuids.SolutionFolderGuid).FirstOrDefault();
+                    if (duplicateProject != null)
+                    {
+                        slnFile.AddSolutionFolders(slnProject, new List<string>() { Path.GetDirectoryName(relativeProjectPath) });
```
Comment (nagilson) [inline src/Cli/dotnet/SlnFileExtensions.cs:115]:
```suggestion
                       // Try making a new folder for the project to put it under so we can still add it despite there being one with the same name already in the parent folder
                        slnFile.AddSolutionFolders(slnProject, new List<string>() { Path.GetDirectoryName(relativeProjectPath) });
```

**Review by nagilson — APPROVED** [review]
LGTM. :D Thanks for responding to my feedback Marc :shipit:

File: src/Cli/dotnet/SlnFileExtensions.cs [hunk 7]
```diff
@@ -255,6 +293,19 @@ private static void AddSolutionFolders(this SlnFile slnFile, SlnProject slnProje
             }
         }
 
+        private static string GetSolutionFolderWithDuplicateProject(SlnFile slnFile, SlnProject slnProject, IList<string> solutionFolders, SlnSection nestedProjectsSection)
```
Comment (nagilson) [inline src/Cli/dotnet/SlnFileExtensions.cs:296]: I do really like this function. It has a complicated task but it's very clear and concise code for what it is 🎆

**Pull request merged**